components/base/kinematicbase/kinematics.go

```diff
@@ -102,50 +110,95 @@ func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]re
+	if capsuleErr != nil {
+		return capsuleErr
 	}
+	movementErr := make(chan error, 1)
+	defer close(movementErr)
 
-	// this loop polls the error state and issues a corresponding command to move the base to the objective
-	// when the base is within the positional threshold of the goal, exit the loop
-	for err = ctx.Err(); err == nil; err = ctx.Err() {
-		col, err := validRegion.CollidesWith(spatialmath.NewPoint(r3.Vector{X: current[0].Value, Y: current[1].Value}, ""))
-		if err != nil {
-			return err
-		}
-		if !col {
-			return errors.New("base has deviated too far from path")
-		}
+	cancelContext, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-		// get to the x, y location first - note that from the base's perspective +y is forward
-		desiredHeading := math.Atan2(current[1].Value-desired[1].Value, current[0].Value-desired[0].Value)
-		commanded, err := ddk.issueCommand(ctx, current, []referenceframe.Input{desired[0], desired[1], {desiredHeading}})
-		if err != nil {
-			return err
-		}
+	utils.PanicCapturingGo(func() {
+		// this loop polls the error state and issues a corresponding command to move the base to the objective
+		// when the base is within the positional threshold of the goal, exit the loop
+		for err := cancelContext.Err(); err == nil; err = cancelContext.Err() {
+			utils.SelectContextOrWait(ctx, 100*time.Millisecond)
+			col, err := validRegion.CollidesWith(spatialmath.NewPoint(r3.Vector{X: current[0].Value, Y: current[1].Value}, ""))
+			if err != nil {
+				movementErr <- err
+				return
+			}
+			if !col {
+				movementErr <- errors.New("base has deviated too far from path")
+				return
+			}
 
-		if !commanded {
-			// no command to move to the x, y location was issued, correct the heading and then exit
-			if commanded, err := ddk.issueCommand(ctx, current, []referenceframe.Input{current[0], current[1], desired[2]}); err == nil {
-				if !commanded {
-					return nil
+			// get to the x, y location first - note that from the base's perspective +y is forward
+			desiredHeading := math.Atan2(current[1].Value-desired[1].Value, current[0].Value-desired[0].Value)
+			commanded, err := ddk.issueCommand(cancelContext, current, []referenceframe.Input{desired[0], desired[1], {desiredHeading}})
+			if err != nil {
+				movementErr <- err
+				return
+			}
+
+			if !commanded {
+				// no command to move to the x, y location was issued, correct the heading and then exit
+				if commanded, err := ddk.issueCommand(cancelContext, current, []referenceframe.Input{current[0], current[1], desired[2]}); err == nil {
+					if !commanded {
+						movementErr <- nil
+						return
+					}
+				} else {
+					movementErr <- err
+					return
 				}
-			} else {
-				return err
+			}
+
+			current, err = ddk.CurrentInputs(cancelContext)
+			if err != nil {
+				movementErr <- err
+				return
 			}
 		}
+		movementErr <- err
+	})
+
+	// watching for movement timeout
+	lastUpdate := time.Now()
+	var prevInputs []referenceframe.Input
```

Comment:
This has the ability to cause a nil pointer dereference on the first iteration; the only reason it doesn't currently is a coincidence of the L2InputMetric implementation.

This should be initialized to CurrentInputs rather than to nil